src/Hamiltonians/HOCartesianEnergyConserved.jl

```diff
@@ -0,0 +1,399 @@
+box_to_aspect(S) = (S[1] - 1) ./ (S .- 1)
+box_to_aspect(S::NTuple{N,Int}) where {N} = (S[1] - 1) .// (S .- 1)
+
+@inline function mode_to_energy(i, S, aspect)
+    # aspect = box_to_aspect(S)
+    states = CartesianIndices(S)
+    # account for 1-indexing here in case aspect ratio is fractional
+    idx = Tuple(states[i]) .- 1
+    E = dot(idx, aspect)
+    # shift back to 1-indexing
+    return E + 1
+end
+
+"""
+    find_Ebounds(i, j, S)
+
+Find the range of single particle energies that a particle in mode `i` can move to while 
+preserving total energy of particles `i` and `j`, and constrained to box `S`.
+"""
+function find_Ebounds(i, j, S, aspect)
+    
+    E_i = mode_to_energy(i, S, aspect)
+    E_j = mode_to_energy(j, S, aspect)
+
+    # shift back to 1-indexing
+    return extrema(largest_two_point_interval(E_i, E_j, S[1]))..., E_i + E_j
+end
+
+"""
+    OccupiedPairsMap(omm::OccupiedModeMap) <: AbstractVector
+
+Get all pairs of particles in `omm`. Pairs involving multiply-occupied modes 
+are counted once, (including self pairing).
+This is an eager iterator whose elements are a tuple of particle indices.
+
+See [`OccupiedModeMap`](@ref).
+"""
+struct OccupiedPairsMap{N,T} <: AbstractVector{T}
+    pairs::SVector{N,T}
+    length::Int
+end
+
+function OccupiedPairsMap(addr::SingleComponentFockAddress{N}) where {N}
+    omm = OccupiedModeMap(addr)
+    T = eltype(omm)
+    P = N * (N - 1) ÷ 2
+    pairs = MVector{P,Tuple{T,T}}(undef)
+    L = 0
+    for i in eachindex(omm)
+        p_i = omm[i]
+        if p_i.occnum > 1
+            L += 1
+            @inbounds pairs[L] = (p_i, p_i)
+        end
+        for j in 1:i-1
+            p_j = omm[j]
+            L += 1
+            @inbounds pairs[L] = (p_i, p_j)
+        end
+    end
+    
+    return OccupiedPairsMap(SVector(pairs), L)
+end
+
+Base.size(opm::OccupiedPairsMap) = (opm.length,)
+function Base.getindex(opm::OccupiedPairsMap, i)
+    @boundscheck 1 ≤ i ≤ opm.length || throw(BoundsError(opm, i))
+    return opm.pairs[i]
+end
+
+"""
+    HOCartesianEnergyConserved(addr; S, η, g = 1.0, interaction_only = false)
+
+Implements a harmonic oscillator in Cartesian basis with contact interactions.
+```math
+\\hat{H} = \\sum_{i} \\epsilon_i n_i + \\frac{g}{2}\\sum_{ijkl} V_{ijkl} a^†_i a^†_j a_k a_l
+```
+Indices ``i, \\ldots`` are ``D``-tuples for a ``D``-dimensional harmonic oscillator. 
+The energy scale is defined by the first dimension i.e. ``\\hbar \\omega_x`` so that 
+single particle energies are ``\\epsilon_i = (i_x + 1/2) + \\eta_y (i_y+1/2) + \\ldots``.
+The factors ``\\eta_y, \\ldots`` allow for anisotropic trapping geometries and are assumed to 
+be greater than `1` so that ``\\omega_x`` is the smallest trapping frequency.
+Matrix elements ``V_{ijkl}`` are for a contact interaction calculated in this basis using 
+first-order perturbation theory. All states with the same noninteracting energy are connected
+by this interaction.
```

Comment:
I'd like to see some more details given such that readers can understand what exactly is implemented.
```suggestion

The matrix elements ``V_{ijkl}`` are for a contact interaction calculated from the product of
one-dimensional harmonic oscillator eigenstates. An additional restriction to conserve
the non-interacting energy is applied as relevant for first-order degenerate perturbation theory. 
All states with the same non-interacting energy are connected by this interaction.
```math
V_{ijkl} = \\prod_{\\kappa \\elem \\{x,y,z\\}}v_{i_\\kappa j_\\kappa k_\kappa l_\\kappa} 
```
The one-dimensional harmonic oscillator matrix elements ``v_{ijkl}`` are obtained from [`four_oscillator_integral_general`](@ref).
```